File: ML_Model/src/bill_extractor.py
import base64
import os
from google import genai
from google.genai import types
from config import GEMINI_API_KEY

genai.configure(api_key=GEMINI_API_KEY)

def encode_image(image_path):
    """Convert image to base64 encoding for API request."""
    with open(image_path, "rb") as image_file:
        return base64.b64encode(image_file.read()).decode("utf-8")

def extract_bill_details(image_path):
    """Use Gemini API to extract bill details."""
    client = genai.Client(api_key=GEMINI_API_KEY)
    image_data = encode_image(image_path)

    prompt = """
    Extract the following details from the bill image:
    - Vendor Name
    - Bill Date
    - Total Amount
    - GST Details (if available)
    - Itemized List (if available)
    Return the data in JSON format.
    """

    contents = [
        types.Content(
            role="user",
            parts=[
                types.Part.from_text(text=prompt),
                types.Part.from_data(data=base64.b64decode(image_data), mime_type="image/jpeg")
            ],
        )
    ]

    generate_content_config = types.GenerateContentConfig(
        temperature=0.7,
        top_p=0.95,
        top_k=40,
        max_output_tokens=8192,
        response_mime_type="application/json",
    )

    response = client.models.generate_content(
        model="gemini-2.0-flash",
        contents=contents,
        config=generate_content_config,
    )

    return response.text  # The extracted JSON output

File: Expense-Tracker-main/backend/utils/billExtractor.js
import { GoogleGenerativeAI } from "@google/generative-ai";
import fs from "fs/promises";

// Get the API key from environment variables
const apiKey = process.env.GEMINI_API_KEY;
const genAI = new GoogleGenerativeAI(apiKey);

// Initialize the Gemini model
const model = genAI.getGenerativeModel({
    model: "gemini-2.0-flash",
});

const generationConfig = {
    temperature: 1,
    topP: 0.95,
    topK: 40,
    maxOutputTokens: 8192,
    responseMimeType: "text/plain",
};

// Extract bill details using Gemini API
export async function extractBill(filePath) {
    try {
        // Read the file and convert to base64
        const imageBuffer = await fs.readFile(filePath);
        const imageBase64 = imageBuffer.toString("base64");

        // Correct API call using generateContent()
        const result = await model.generateContent({
            contents: [
                {
                    role: "user",
                    parts: [
                        {
                            inlineData: {
                                mimeType: "image/jpeg",
                                data: imageBase64,
                            },
                        },
                        {
                            text: "Extract the following details from the bill image:\n" +
                                "- Vendor Name\n" +
                                "- Type of the Expense\n" +
                                "- Category of the Expense\n" +
                                "- Bill Date\n" +
                                "- Total Amount\n" +
                                "Return the data in valid JSON format.",
                        },
                    ],
                },
            ],
            generationConfig,
        });

        // ✅ Correct way to get the response text
        console.log("Raw result:", JSON.stringify(result, null, 2));

        // Extract text response correctly
        const responseText = result.response.candidates[0].content.parts[0].text.trim();

        // ✅ Remove backticks and parse clean JSON
        const cleanedResponse = responseText.replace(/```json\n|```/g, "").trim();


        console.log("Extracted Bill Data:", cleanedResponse);
        return cleanedResponse;
    } catch (error) {
        console.error("Error while extracting bill:", error.message);
        throw error;
    }
}
